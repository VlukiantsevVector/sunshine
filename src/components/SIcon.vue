<template>
  <span
    :class="classes()"
    v-on="$listeners"
  ><span
    :style="iconStyle"
    :class="iconClass"
  />
  </span>
</template>

<script lang="ts">
import Vue from 'vue';
import mixins from 'vue-typed-mixins';
import ClassesMixin from './internal/ClassesMixin';

import '../icons/icon54/style.css';
import '../icons/sunshine24/css/sunshine24-embedded.css';


export default mixins(ClassesMixin).extend({
  name: 's-icon',

  props: {
    package: {
      type: String,
      default: 'icon',
    },

    name: {
      type: String,
      required: true,
    },

    rotate: {
      type: Number,
      default: 0,
    },
  },

  computed: {
    iconClass(): string {
      return `${this.package}-${this.name}`;
    },

    iconStyle(): any {
      return {
        display: 'inline-block',
        transform: `rotate(${this.rotate}deg)`,
      };
    },
  },
});
</script>
