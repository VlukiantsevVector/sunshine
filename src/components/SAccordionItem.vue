<template>
  <div :class="$class()">
    <a
      href="#"
      :class="$class('heading')"
      @_click="$emit('click')"
      @click="onClick"
    >
      <span :class="$class('title')">
        <slot name="heading">
          {{ heading }}
        </slot>
      </span>
      <s-icon
        v-if="invalid"
        package="sunshine24"
        name="warning"
        :class="$class('warning')"
      />
      <s-icon
        package="sunshine24"
        name="arrow"
        :class="$class('chevron', chevronOptions)"
      />
    </a>

    <div
      v-show="showBody"
      :class="$class('body', bodyOptions)"
      :style="bodyStyle"
      @transitionend="onTransitionEnd"
    >
      <div
        ref="inner"
        :class="$class('content')"
      >
        <slot />
      </div>
    </div>
  </div>
</template>

<script>
import Vue from 'vue';
import SIcon from './SIcon.vue';

export default Vue.extend({
  name: 'SAccordionItem',

  inject: ['$validator'],

  components: {
    SIcon,
  },

  props: {
    heading: {
      type: String,
      default: '',
    },

    invalid: {
      type: Boolean,
      default: false,
    },

    defaultOpen: {
      type: Boolean,
      default: false,
    },
  },

  data() {
    return {
      open: this.defaultOpen,
      height: 0,
      transitioning: false,
    };
  },

  watch: {
    async open(val) {
      if (val) {
        this.$parent.notifyOpen(this);
      }

      // Make sure height is defined at the beginning of the animation
      if (!val) {
        this.height = this.$refs.inner.offsetHeight;
      }

      this.transitioning = true;

      // Need to wait for height to be correct
      await Vue.nextTick();

      // We read offsetHeight from the element, even in the case when
      // we're setting height to 0, because accessing this property will
      // actually force the browser to do a 'reflow', making sure the height
      // is set to the initial value, before we change it. If we don't do this,
      // the transition may not always occur.
      // Ref: https://stackoverflow.com/a/16575811/137627
      const height = this.$refs.inner.offsetHeight;

      if (val) {
        this.height = height;
      } else {
        this.height = 0;
      }
    },
  },

  computed: {
    showBody() {
      return this.open || this.transitioning;
    },

    chevronOptions() {
      return { rotate: !this.open };
    },

    bodyOptions() {
      return {
        open: this.open,
        transitioning: this.transitioning,
      };
    },

    bodyStyle() {
      if (!this.transitioning) return {};

      return {
        height: `${this.height}px`,
        opacity: this.height === 0 ? 0 : 1,
      };
    },
  },

  methods: {
    onClick() {
      this.open = !this.open;
    },

    onTransitionEnd() {
      this.transitioning = false;
    },

    close() {
      this.open = false;
    },
  },

  mounted() {
    this.$parent.registerItem(this);
  },
});
</script>
