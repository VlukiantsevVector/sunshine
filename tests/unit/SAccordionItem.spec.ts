import { mount } from '@vue/test-utils';
import SAccordionItem from '@/components/SAccordionItem.vue';
import SIcon from '@/components/SIcon.vue';

let wrapper : any;

// Creating a stub for SAccordion.
beforeAll(() => {
  const parent = {
    components: {
      SIcon,
    },
    methods: {
      registerItem(item : any) {},
    },
  };

  const item = '<h1>Test</h1>';
  wrapper = mount(SAccordionItem, {
    parentComponent: parent,
    slots: {
      heading: 'This is a header',
      default: item,
    },
  });
});

describe('SAccordionItem', () => {
  it('renders text in header and default slot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('show content of inner slot when clicked', async () => {
    wrapper.find('a').trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.contains('.s-accordion-item__body--open')).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('hides content of inner slot when clicked again', async () => {
    wrapper.find('a').trigger('click');
    await wrapper.vm.$nextTick();
    // expect(wrapper.find('display')).toEqual('none');
    //
    // display shows "". In the snapshot the class is s-accordion-item__body--transitioning"
    // and style="height: 0px; opacity: 0;"
    // Seems like we're getting a snapshot of the transition between open -> close

    expect(wrapper.html()).toMatchSnapshot();
  });
});
