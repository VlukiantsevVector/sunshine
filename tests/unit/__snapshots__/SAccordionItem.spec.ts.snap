// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SAccordionItem hides content of inner slot when clicked again 1`] = `
<div class="s-accordion-item__wrapper">
  <div class="s-accordion-item__status-wrapper">
    <div class="s-accordion-item__status"></div>
  </div>
  <div class="s-accordion-item"><a href="#" class="s-accordion-item__heading"><span class="s-accordion-item__title">This is a header</span> <span class="s-icon s-accordion-item__chevron s-accordion-item__chevron--rotate"><span class="sunshine24-arrow"></span></span></a>
    <div class="s-accordion-item__body s-accordion-item__body--transitioning" style="height: 0px; opacity: 0;">
      <div class="s-accordion-item__content">
        <h1>Test</h1>
      </div>
    </div>
  </div>
</div>
`;

exports[`SAccordionItem renders text in header and default slot 1`] = `
<div class="s-accordion-item__wrapper">
  <div class="s-accordion-item__status-wrapper">
    <div class="s-accordion-item__status"></div>
  </div>
  <div class="s-accordion-item"><a href="#" class="s-accordion-item__heading"><span class="s-accordion-item__title">This is a header</span> <span class="s-icon s-accordion-item__chevron s-accordion-item__chevron--rotate"><span class="sunshine24-arrow"></span></span></a>
    <div class="s-accordion-item__body" style="display: none;">
      <div class="s-accordion-item__content">
        <h1>Test</h1>
      </div>
    </div>
  </div>
</div>
`;

exports[`SAccordionItem show content of inner slot when clicked 1`] = `
<div class="s-accordion-item__wrapper">
  <div class="s-accordion-item__status-wrapper">
    <div class="s-accordion-item__status s-accordion-item__status--open"></div>
  </div>
  <div class="s-accordion-item"><a href="#" class="s-accordion-item__heading"><span class="s-accordion-item__title s-accordion-item__title--open">This is a header</span> <span class="s-icon s-accordion-item__chevron"><span class="sunshine24-arrow"></span></span></a>
    <div class="s-accordion-item__body s-accordion-item__body--open" style="">
      <div class="s-accordion-item__content">
        <h1>Test</h1>
      </div>
    </div>
  </div>
</div>
`;
